source/application/Utils/File.cpp

```diff
@@ -0,0 +1,357 @@
+/*
+ * Copyright (c) 2016- Hourglass Resurrection Team
+ * Hourglass Resurrection is licensed under GPL v2.
+ * Refer to the file COPYING.txt in the project root.
+ */
+
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+
+#include <algorithm>
+
+#include "application/logging.h"
+#include "File.h"
+
+Utils::File::File::File(const std::wstring& filename, DWORD access, DWORD share, DWORD create, DWORD flags) :
+    m_file(INVALID_HANDLE_VALUE)
+{
+    m_file = CreateFileW(filename.c_str(), access, share, nullptr, create, flags, nullptr);
+    if (m_file == INVALID_HANDLE_VALUE)
+    {
+        PrintLastError(L"CreateFileW", GetLastError());
+    }
+}
+
+Utils::File::File::~File()
+{
+    CloseFile();
+}
+
+bool Utils::File::File::ReadFile(LPBYTE buffer, SIZE_T length, SIZE_T* read)
+{
+    if (!IsValid())
+    {
+        return false;
+    }
+
+    SIZE_T read_bytes = 0;
+    SIZE_T length_remains = length;
+    while (read_bytes < length)
+    {
+        DWORD read_length = static_cast<DWORD>(std::min<SIZE_T>(static_cast<SIZE_T>(MAXDWORD), length_remains));
+        DWORD bytes = 0;
+        BOOL rv = ::ReadFile(m_file, buffer, read_length, &bytes, nullptr);
+        if (rv == FALSE)
+        {
+            return false;
+        }
+        if (bytes == 0)
+        {
+            break;
+        }
+        read_bytes += bytes;
+        length_remains -= bytes;
+    }
+            
+    if (read != nullptr)
+    {
+        *read = read_bytes;
+    }
+    return true;
+}
+
+bool Utils::File::File::WriteFile(const LPBYTE buffer, SIZE_T length, SIZE_T * written)
+{
+    // TODO
+    return false;
+}
+
+bool Utils::File::File::CloseFile()
+{
+    if (!IsValid())
+    {
+        return false;
+    }
+    auto file = m_file;
+    m_file = INVALID_HANDLE_VALUE;
+    return (CloseHandle(file) == TRUE);
+}
+
+SIZE_T Utils::File::File::GetSize()
+{
+    if (!IsValid())
+    {
+        return 0;
+    }
+    LARGE_INTEGER size;
+    memset(&size, 0, sizeof(size));
+
+    if (GetFileSizeEx(m_file, &size) != TRUE)
+    {
+        return 0;
+    }
+    return size.QuadPart;
+}
+
+LONGLONG Utils::File::File::Seek(LONGLONG distance, DWORD starting_point)
+{
+    if (!IsValid())
+    {
+        return -1;
+    }
+    LARGE_INTEGER move;
+    PLONG high_part = &move.HighPart;
+    move.QuadPart = distance;
+    if (move.HighPart == 0)
+    {
+        high_part = nullptr;
+    }

```

Comment:
SetFilePointer requires this. If you're going to seek a distance that fits within 32 bits, the `high_part` parameter must be null.